src/test/register_api_test.cc

```diff
@@ -0,0 +1,374 @@
+  }
+
+  TRITONSERVER_Server* server_ = nullptr;
+};
+
+TEST_F(RegisterApiTest, Register)
+{
+  // Request to load "model_0" which should fail
+  FAIL_TEST_IF_NOT_ERR(
+      TRITONSERVER_ServerLoadModel(server_, "model_0"),
+      TRITONSERVER_ERROR_INTERNAL,
+      "failed to load 'model_0', failed to poll from model repository",
+      "loading model 'model_0'");
+
+  // Registering a repository "models_0" where contains "model_0"
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_ServerRegisterModelRepository(
+          server_, "models_0", nullptr, 0),
+      "registering model repository 'models_0'");
+  // Request to load "model_0"
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_ServerLoadModel(server_, "model_0"),
+      "loading model 'model_0'");
+}
+
+TEST_F(RegisterApiTest, RegisterWithMap)
+{
+  // Registering a repository "models_0" where contains "model_0", but with
+  // different name mapping
+  const char* override_name = "name_0";
+  std::shared_ptr<TRITONSERVER_Parameter> managed_param(
+      TRITONSERVER_ParameterNew(
+          "model_0", TRITONSERVER_PARAMETER_STRING, override_name),
+      TRITONSERVER_ParameterDelete);
+  ASSERT_TRUE(managed_param != nullptr) << "failed to create name mapping pair";
+  std::vector<const TRITONSERVER_Parameter*> name_map{managed_param.get()};
+
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_ServerRegisterModelRepository(
+          server_, "models_0", name_map.data(), name_map.size()),
+      "registering model repository 'models_0'");
+
+  // Request to load "model_0" which should fail
+  FAIL_TEST_IF_NOT_ERR(
+      TRITONSERVER_ServerLoadModel(server_, "model_0"),
+      TRITONSERVER_ERROR_INTERNAL,
+      "failed to load 'model_0', failed to poll from model repository",
+      "loading model 'model_0'");
+  // Request to load "name_0"
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_ServerLoadModel(server_, "name_0"),
+      "loading model 'name_0'");
+}
+
+TEST_F(RegisterApiTest, RegisterTwice)
+{
+  // Registering a startup repository
+  FAIL_TEST_IF_NOT_ERR(
+      TRITONSERVER_ServerRegisterModelRepository(
+          server_, "empty_models", nullptr, 0),
+      TRITONSERVER_ERROR_ALREADY_EXISTS,
+      "model repository 'empty_models' has been registered",
+      "registering model repository 'empty_models'");
+}
+
+TEST_F(RegisterApiTest, RegisterTwice2)
+{
+  // Registering the same repository twice
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_ServerRegisterModelRepository(
+          server_, "models_0", nullptr, 0),
+      "registering model repository 'models_0'");
+
+  FAIL_TEST_IF_NOT_ERR(
+      TRITONSERVER_ServerRegisterModelRepository(
+          server_, "models_0", nullptr, 0),
+      TRITONSERVER_ERROR_ALREADY_EXISTS,
+      "model repository 'models_0' has been registered",
+      "registering model repository 'models_0'");
+}
+
+TEST_F(RegisterApiTest, RegisterWithMultiMap)
+{
+  // Registering a repository "models_0" where contains "model_0",
+  // and "model_0" is mapped to two different names
```

Comment:
@GuanLuo @deadeyegoodwin Is this a case we want/need to support?

I can change the mapping values to hold a list of model names, but I feel that it'd introduce complexity for an edge case that seems to have very limited utility or use. Do users need the ability to map the same subdirectory to multiple model names?

Reply:
I agree. Let's enforce that there can be only a single name for any model. There needs to be some testing around this.